_posts/en/newsletters/2019-05-14-newsletter.md

```diff
@@ -0,0 +1,648 @@
+  with new sighash types or other changes.
+
+- **No activation mechanism specified:** if users decide they want to
+  begin enforcing the soft fork's new rules, safety requires that enough
+  of them begin enforcing the new rules at the same block so that miners
+  are deterred from creating blocks that violate the new rules.  Various
+  mechanisms have been used to accomplish this in the past and other
+  options have been described for potential future use.  However,
+  bip-taproot doesn't mention any of these techniques.  Optech agrees
+  with the primary author of the BIP that [activation discussion is
+  premature][discuss activation].  We first need to ensure that there's
+  widespread agreement that the proposals are safe and desirable before
+  we start a debate about the best way to activate them.
+
+### Single-sig spending using Taproot
+
+To look at what's in the proposals, we'll primarily examine how existing
+use cases could be transferred to Taproot.  The best place to start is by
+looking at the way most wealth is transferred via the Bitcoin protocol
+right now: single-sig P2PKH and P2WPKH spends.
+
+Single-sig P2WPKH wallets today currently generate a private key, derive
+a public key (pubkey) from it, and hash that pubkey to create the
+witness program for a bech32 address.  (P2PKH is functionally identical
+but uses a different script and a different address encoding.)
+
+| Object      | Operation                                                       | Example result    |
+|-|-|-|
+| Private key | read 32 bytes from [CSPRNG][], or using [BIP32][] HD derivation | `0x807d[...]0101` |
+| Public key  | point(0x807d[...]0101), or using [BIP32][] HD public derivation | `0x02e5[...]3c23` |
+| Hash        | ripemd(sha256(0x0202e5[...]3c23))                               | `0x006e[...]05d6` |
+| Address     | bech32.encode('bc', 0, 0x006e[...]05d6)                         | `bc1qd6[...]24zh` |
+
+Under Taproot, the hashing step will be skipped and so the bech32
+address will contain the pubkey directly, with one small change.
+Currently, 33-byte Bitcoin-style pubkeys are encoded to start with either
+a 0x02 or 0x03 to allow validators to reconstruct the key's Y-coordinate
+on the secp256k1 elliptic curve; in bip-taproot, the value of this byte
+is reduced by two so that 0x02 becomes 0x00 and 0x03 becomes 0x01.  The
+meaning stays the same but using the low bit for the values frees up the
+remaining bits for future soft forks.
+Also the witness version is changed from the `0` used
+for P2WPKH/P2WSH to a `1`.
+
+| Object           | Operation                               | Example result    |
+|-|-|-|
+| Private key      | (Same as above)                         | `0x807d[...]0101` |
+| Public key       | (Same as above)                         | `0x02e5[...]3c23` |
+| Alter key prefix | (key[0] - 2),key[1:33])                 | `0x00e5[...]3c23` |
+| Address          | bech32.encode('bc', 1, 0x00e5[...]3c23) | `bc1pqr[...]xg73` |
```

Comment:
What about including the full private key `0x807d[...]0101` to allow for using all the examples in the post as test vectors? I think that would be helpful.

Reply:
@practicalswift test vectors are good, but I don't want to encourage people to write implementations based on an overview of the proposals.  If anyone wants to write a toy implementation now, I'd highly encourage them to try to read only the proposed BIPs; that way they can identify any confusing sections or omitted details or other problems in what will be the primary sources for Taproot and Tapscript details---and they can then ask the BIP author to make edits while the document is still a discussion draft.